### DISCUSSION TITLE
A bug in PlatformIO.ini [env:demo] and [env:m5plusdemo]?

### DISCUSSION THREAD

**Discussion by airnocker** [PlummersSoftwareLLC/NightDriverStrip#433] [Q&A]
My MCU is the M5StickC Plus.  Setting the PlatformIO.ini's **default_env = laserline** works like a charm.
Setting default_env = demo fails with a "Can't open COM8; port doesn't exist" or some such error.  It doesn't seem to know I'm using the M5StickC Plus.

However, if I modify the [env:demo] section to this:

[env:demo]
extends         = dev_m5stick-c-plus                ;originally dev_esp32
build_flags     = -DDEMO=1
                          -Ofast
                          ${dev_m5stick-c-plus.build_flags}

It compiles just fine.

I tried specifying the default_env = m5plusdemo in the .ini and leaving the [env:demo] section as originally written and this also does not compile.

Both project builds sometimes errors on not finding COM8 automatically and I have to manually equate it as upload_port = COM8
But then again, somethimes this manual equate results in the same "Can't open COM8" error.

Suggestion on this behavior are welcomed.

**Comment by rbergen** [comment]
What exactly are the errors you're getting when you try to build m5plusdemo? The CI on this project builds every project with every merge, so we know that the code as-is _can_ build. 

Your repeated mention of COM8 makes me think the actual compile may complete, but the upload to the stick fails. Which still leaves you without it running the project you're trying to put onto it, but leans towards a different avenue of investigation.

**Comment by rbergen** [comment]
Having let this simmer for a bit, I believe the "Can't open COM8" error can be caused by the COM port already being in use. Do you ever happen to use the "Upload and Monitor" or "Monitor" features within VSCode/PIO?  
I know that these can "linger" (stay open) if you switch projects in platformio.ini, which can then cause problems if you try to (build and) upload another project.

**Reply by airnocker** [reply]
Hi @rbergen.  You would know better than I.  But this would seem to be a "peculiarity" of VSCode/PIO because I do not experience this with the Arduino IDE.

Allow me to clarify when this happens.  Lets say I've just compiled for the laserline environment and upload it successfully to my M5StickC Plus.  The M5 is displaying the colorbar VU meter and loud sounds result in the laser cannon burst down the strip.  5, 10, 15 minutes later I change to the demo environment and compile it.  Compile fails with the error mentioned.  

So, your experience is that the VSCode/PIO "system" can cause the COM* link to stay open after an upload to the device on COM* has completed, as though it is still in use, resulting in any attempt to write (upload) the new code to the "device" to fail. Hmmm

**Reply by rbergen** [reply]
I obviously can't see exactly what happens on your VSCode/PIO setup, but I can share some of mine. This is a screen snippet of the bottom-right part of my VSCode window:
![Screenshot 2023-09-20 080753](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/10425218/55b668dd-1e1f-4194-9c15-9412dffb88c6)

The panel at the right shows that I consecutively:

- Executed a build of the "spectrum" project
- Performed an "Upload and Monitor" of the build to my M5StickC Plus, which is still running, as indicated by the "spinning wheel" right next to that line.
- Switched to the "m5plusdemo" project and built that.

So far so good.

If I then try to "Upload and monitor" the "m5plusdemo" build while the "Upload and Monitor" of "spectrum" is still active, I get this:
![Screenshot 2023-09-20 080859](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/10425218/7baec470-977a-41ef-9b0d-1a8b16cbcd05)

As you can see, it gives me the same "Could not open COMn, the port doesn't exist" message you've been mentioning.

(What the snippets also show is that I select the project **and** the COM port in the status bar at the very bottom of the VSCode window. That way I can tell VSCode/PIO what I want to build and how I want to upload it, without ever having to "hard code" it into platformio.ini.)

Again, I can't say with any certainty if this is exactly what you're running into, but maybe it can give you some pointers for further investigation.

**Reply by airnocker** [reply]
Thanks @rbergen, that is exactly what my "success" and "failed" screens look like WITH ONE EXCEPTION DESCRIBED BELOW.  This should help me better understand how the PIO works, what feedback the PIO is giving me in these areas on the screen and how to interpret them.

The exception is the below the Terminal window mine looks like this:
![Bottom screen](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/31143220/cc96b595-f1a8-44d7-80ef-8d5b6a70de64)

**Reply by airnocker** [reply]
I just noticed last night that I could click on "Auto" then select from the pop-down list my COM port.  Maybe that's what I've not being doing or paying attention to.  Right now 'demo' fails whether the PIO status bar show "Auto" or "COM8".  I found this explanation by clicking on the red "Platform...X":

![demo-FAILED](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/31143220/183936a5-c5b9-4cd1-9b54-6c46704ca2b5)

**Reply by rbergen** [reply]
I think the clue for analyzing this one is in the terminal lines above the `*** [upload] Error 2`. Does that panel allow you to scroll up and inspect the earlier lines in its output?

**Reply by airnocker** [reply]
I've just learned how to clean-up the PIO status bar items.

demo, which compiled with SUCCESS last night, today reflected the error as we've discussed.  Unplugging the USB cable to the M5StickC Plus and reinserting it resulted with a build and upload SUCCESS shown here.  But my project folder isn't showing 'demo' or 'laserline' as yours does but rather 'Default (NightDriverStrip)'.  It a minor thing to me but a curiosity for why.

![demo-SUCCESS](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/31143220/8d01fb9c-2627-4cd8-b253-93811f5a209f)

**Reply by rbergen** [reply]
Based on earlier screen snippets you shared I'm pretty sure it's not "old" PlatformIO tasks hanging on to the COM port when you get the error, but something is. That "something" apparently gives up when the connection is physically broken. I'll be honest: I don't know how I could analyze this further without being behind your screen.

...maybe the disconnect/reconnect is the "pragmatic workaround" to adopt, at least for the time being.

**Comment by rbergen** [comment]
> The exception is the below the Terminal window mine looks like this:
> ![Bottom screen](https://user-images.githubusercontent.com/31143220/269381613-cc96b595-f1a8-44d7-80ef-8d5b6a70de64.jpg)

That's interesting. That's not a status bar provided by PlatformIO, but one provided by a CMake plugin. I haven't yet had the need to figure out how to prioritize one over another. I'll see if I can dig up some more info on this particular point.

**Reply by airnocker** [reply]
I just right-clicked on the status bar and unchecked CMAKE TOOLS (Extension) to make it go away.

**Reply by rbergen** [reply]
Haha, I can only say that that's a very good thing to try early! It's 9:18 PM here and I'm about to have dinner, so please forgive me for being a little slow right this moment. :)

**Reply by airnocker** [reply]
I am in no hurry and do not want to impose any sense of urgency in my "learning quest".

I changed the ini default_envs to laserline and got the same FAILED result.  Here is the screenshot of what is immediately above the [upload] Error 2:

![laserline-FAILED](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/31143220/089d1aea-de25-48e8-98b6-e5759ced910b)

**Reply by airnocker** [reply]
Update for: NightDriverStrip (local)
I restored PlatformIO.ini lines 281-286 to their original state, saved the .ini file.
Specified "m5plusdemo" on line 22 of the .ini, saved the .ini file.
Status bar shows "COM8"
Ran build and upload: failed Error 2: Could not open COM8
In PIO status bar, clicked "Set upload/monitor/test port" and chose:

![COM8-text](https://github.com/PlummersSoftwareLLC/NightDriverStrip/assets/31143220/cac30082-94eb-4fe7-bc5f-c974e1b3dfb7)

Ran build and upload: SUCCESS

I'm not at all sure why this worked but I'm glad it did.

**Reply by rbergen** [reply]
That's great news!

**Reply by airnocker** [reply]
Small wonders never cease, eh?
So, I'm finding the README.md has many answers to my questions.
Thanks for your patience.

**Reply by rbergen** [reply]
You're welcome, and the README.md thing is also good news. When you write the documentation, you never really know how much use people will get out of it. :)
